compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeResolver.java

```diff
@@ -0,0 +1,2212 @@
+                break;
+            case USER_DEFINED_TYPE:
+                resultType = resolveTypeDesc((BLangUserDefinedType) td, data);
+                break;
+            case BUILT_IN_REF_TYPE:
+                resultType = resolveTypeDesc((BLangBuiltInRefTypeNode) td, symEnv);
+                break;
+            case FINITE_TYPE_NODE:
+                resultType = resolveSingletonType((BLangFiniteTypeNode) td, symEnv);
+                break;
+            case TABLE_TYPE:
+                resultType = resolveTypeDesc((BLangTableTypeNode) td, data);
+                break;
+            case STREAM_TYPE:
+                resultType = resolveTypeDesc((BLangStreamType) td, data);
+                break;
+            default:
+                throw new AssertionError("Invalid type");
+        }
+
+        BType refType = Types.getReferredType(resultType);
+        if (refType != symTable.noType) {
+            // If the typeNode.nullable is true then convert the resultType to a union type
+            // if it is not already a union type, JSON type, or any type
+            if (td.nullable && resultType.tag == TypeTags.UNION) {
+                BUnionType unionType = (BUnionType) refType;
+                unionType.add(symTable.nilType);
+            }
+        }
+
+        symResolver.validateDistinctType(td, resultType);
+        if (td.getBType() == null) {
+            td.setBType(resultType);
+        }
+        data.env = prevEnv;
+        data.typeDefinition = prevDefn;
+        data.depth = prevDepth;
+        return resultType;
+    }
+
+    private BType resolveTypeDesc(BLangValueType td, SymbolEnv symEnv) {
+        SymbolResolver.AnalyzerData data = new SymbolResolver.AnalyzerData(symEnv);
+        return visitBuiltInTypeNode(td, data, td.typeKind);
+    }
+
+    private BType resolveTypeDesc(BLangConstrainedType td, ResolverData data) {
+        currentDepth = data.depth;
+        TypeKind typeKind = ((BLangBuiltInRefTypeNode) td.getType()).getTypeKind();
+
+        if (typeKind == TypeKind.MAP) {
+            return resolveMapTypeDesc(td, data);
+        } else if (typeKind == TypeKind.XML) {
+            return resolveXmlTypeDesc(td, data);
+        }
+
+        if (td.getBType() != null) {
+            return td.getBType();
+        }
+
+        BType type = resolveTypeDesc(data.env, data.typeDefinition, data.depth + 1, td.type, data);
+
+        BType constrainedType = symTable.neverType;
+        switch (typeKind) {
+            case FUTURE:
+                constrainedType = new BFutureType(TypeTags.FUTURE, symTable.empty, null);
+                break;
+            case TYPEDESC:
+                constrainedType = new BTypedescType(symTable.empty, null);
+                break;
+        }
+        BTypeSymbol typeSymbol = type.tsymbol;
+        constrainedType.tsymbol = Symbols.createTypeSymbol(typeSymbol.tag, typeSymbol.flags, typeSymbol.name,
```

Comment:
Isn't this Update the neverType's symbol? We may have to rewrite this without using the two switch statement pattern.